File: commons-core/src/main/java/xyz/wagyourtail/commons/core/reader/CharReader.java
package xyz.wagyourtail.commons.core.reader;

import lombok.val;
import org.jetbrains.annotations.Nullable;
import xyz.wagyourtail.commons.core.StringUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public abstract class CharReader<T extends CharReader<? super T>> {

    public static final CharAccepter WHITESPACE = new CharAccepter() {
        @Override
        public boolean accept(char c) {
            return Character.isWhitespace(c);
        }
    };
    public static final CharAccepter NEWLINE = new CharAccepter() {
        @Override
        public boolean accept(char c) {
            return c == '\n';
        }
    };
    public static final CharAccepter COMMA = new CharAccepter() {
        @Override
        public boolean accept(char c) {
            return c == ',';
        }
    };
    public static final CharAccepter NOT_NEWLINE_WHITESPACE = and(WHITESPACE, not(NEWLINE));

    public CharReader() {
    }

    public static CharAccepter and(final CharAccepter a, final CharAccepter b) {
        Objects.requireNonNull(a);
        Objects.requireNonNull(b);
        return new CharAccepter() {
            @Override
            public boolean accept(char c) {
                return a.accept(c) && b.accept(c);
            }
        };
    }

    public static CharAccepter or(final CharAccepter a, final CharAccepter b) {
        Objects.requireNonNull(a);
        Objects.requireNonNull(b);
        return new CharAccepter() {
            @Override
            public boolean accept(char c) {
                return a.accept(c) || b.accept(c);
            }
        };
    }

    public static CharAccepter not(final CharAccepter a) {
        Objects.requireNonNull(a);
        return new CharAccepter() {
            @Override
            public boolean accept(char c) {
                return !a.accept(c);
            }
        };
    }

    public static CharAccepter of(final char character) {
        return new CharAccepter() {
            @Override
            public boolean accept(char c) {
                return character == c;
            }
        };
    }

    /**
     * @return either a char, or -1 for eof
     */
    public abstract int peek();

    /**
     * @return either a char, or -1 for eof
     */
    public abstract int take();

    /**
     * @param count the number of chars to skip
     * @return the number of chars skipped
     */
    public int skip(int count) {
        for (int i = 0; i < count; i++) {
            int ch = take();
            if (ch == -1) {
                return i;
            }
        }
        return count;
    }

    /**
     * @return a string of the next count chars
     */
    public String take(int count) {
        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < count; i++) {
            int ch = take();
            if (ch == -1) break;
            builder.append((char) ch);
        }
        return builder.toString();
    }

    /**
     * @return a copy of the reader at the current position
     */
    @Deprecated
    public T copy() {
        return copy(Integer.MAX_VALUE);
    }

    /**
     * @return a copy of the reader at the current position, with a limit to the number of characters it can read.
     * @since 1.0.4
     */
    @Deprecated
    public abstract T copy(int limit);

    /**
     * set this position as remembered.
     */
    public void mark() {
        mark(Integer.MAX_VALUE);
    }

    public abstract void mark(int limit);

    /**
     * reset to remembered mark position, or beginning if no mark.
     */
    public abstract void reset();

    public boolean exhausted() {
        return peek() == -1;
    }

    public String takeRemaining() {
        StringBuilder sb = new StringBuilder();
        while (peek() != -1) {
            sb.append((char) take());
        }
        return sb.toString();
    }

    public String takeLine() {
        return takeUntil('\n');
    }

    public String takeUntil(char character) {
        StringBuilder sb = new StringBuilder();
        int next = peek();
        while (next != -1 && next != character) {
            sb.append((char) take());
            next = peek();
        }
        return sb.toString();
    }

    public String takeUntil(String characters) {
        StringBuilder sb = new StringBuilder();
        int next = peek();
        String taken = "";
        while (next != -1) {
            do {
                if (characters.startsWith(taken + (char) next)) {
                    if (taken.isEmpty()) {
                        mark();
                    }
                    taken += (char) next;
                    break;
                } else if (!taken.isEmpty()) {
                    val newStart = taken.indexOf(characters.charAt(0), 1);
                    if (newStart == -1) {
                        taken = "";
                        if (characters.startsWith(taken + (char) next)) {
                            mark();
                            taken += (char) next;
                        }
                        break;
                    }
                    reset();
                    skip(newStart);
                    mark();
                    taken = taken.substring(newStart);
                    skip(taken.length());
                }
            } while (!taken.isEmpty());
            sb.append((char) take());
            if (characters.equals(taken)) {
                val out = StringUtils.removeSuffix(sb.toString(), characters);
                reset();
                return out;
            }
            next = peek();
        }
        return sb.toString();
    }

    public String takeUntil(CharAccepter accepter) {
        Objects.requireNonNull(accepter);
        StringBuilder sb = new StringBuilder();
        int next = peek();
        while (next != -1 && !accepter.accept((char) next)) {
            sb.append((char) take());
            next = peek();
        }
        return sb.toString();
    }

    public String takeWhile(CharAccepter accepter) {
        Objects.requireNonNull(accepter);
        StringBuilder sb = new StringBuilder();
        int next = peek();
        while (next != -1 && accepter.accept((char) next)) {
            sb.append((char) take());
            next = peek();
        }
        return sb.toString();
    }

    public String takeWhitespace() {
        return takeWhile(WHITESPACE);
    }

    public String takeNonNewlineWhitespace() {
        return takeWhile(NOT_NEWLINE_WHITESPACE);
    }

    public @Nullable String takeNext() {
        return takeNext(WHITESPACE);
    }

    public @Nullable String takeNext(char sep) {
        int next = peek();
        if (next == -1 || next == '\n') return null;
        String value;
        if (next == '"') value = takeString();
        else {
            StringBuilder sb = new StringBuilder();
            while (next != -1) {
                if (next == '\n' || next == sep) break;
                sb.append((char) take());
                next = peek();
            }
            value = sb.toString();
        }
        // take trailing sep
        next = peek();
        if (next == sep) {
            take();
        }
        return value;
    }

    public @Nullable String takeNext(CharAccepter sep) {
        int next = peek();
        if (next == -1 || next == '\n') return null;
        String value;
        if (next == '"') value = takeString();
        else value = takeUntil(or(sep, NEWLINE));
        // take trailing sep
        next = peek();
        if (next != -1 && next != '\n' && sep.accept((char) next)) {
            take();
        }
        return value;
    }

    public @Nullable String takeNextLiteral() {
        return takeNextLiteral(WHITESPACE);
    }

    public @Nullable String takeNextLiteral(char sep) {
        int next = peek();
        if (next == -1 || next == '\n') return null;
        StringBuilder sb = new StringBuilder();
        while (next != -1) {
            if (next == '\n' || next == sep) break;
            sb.append((char) take());
            next = peek();
        }
        // take trailing sep
        if (next == sep) {
            take();
        }
        return sb.toString();
    }

    public @Nullable String takeNextLiteral(CharAccepter sep) {
        int next = peek();
        if (next == -1 || next == '\n') return null;
        StringBuilder sb = new StringBuilder();
        while (next != -1) {
            if (next == '\n' || sep.accept((char) next)) break;
            sb.append((char) take());
            next = peek();
        }
        // take trailing sep
        if (next != -1 && next != '\n' && sep.accept((char) next)) {
            take();
        }
        return sb.toString();
    }

    public List<String> takeRemainingOnLine() {
        return takeRemainingOnLine(WHITESPACE);
    }

    public List<String> takeRemainingOnLine(char sep) {
        List<String> args = new ArrayList<>();
        String next = takeNext(sep);
        while (next != null) {
            args.add(next);
            next = takeNext(sep);
        }
        return args;
    }

    public List<String> takeRemainingOnLine(CharAccepter sep) {
        List<String> args = new ArrayList<>();
        String next = takeNext(sep);
        while (next != null) {
            args.add(next);
            next = takeNext(sep);
        }
        return args;
    }

    public List<String> takeRemainingLiteralOnLine() {
        return takeRemainingOnLine(WHITESPACE);
    }

    public List<String> takeRemainingLiteralOnLine(char sep) {
        List<String> args = new ArrayList<>();
        String next = takeNextLiteral(sep);
        while (next != null) {
            args.add(next);
            next = takeNextLiteral(sep);
        }
        return args;
    }

    public List<String> takeRemainingLiteralOnLine(CharAccepter sep) {
        List<String> args = new ArrayList<>();
        String next = takeNextLiteral(sep);
        while (next != null) {
            args.add(next);
            next = takeNextLiteral(sep);
        }
        return args;
    }

    public static final int TAKE_STRING_LEINIENT             = 0b1;
    public static final int TAKE_STRING_ESCAPE_DOUBLE_QUOTE  = 0b10;
    public static final int TAKE_STRING_ESCAPE_NEWLINE       = 0b100;
    public static final int TAKE_STRING_MULTILINE            = 0b1000;
    public static final int TAKE_STRING_NO_START_QUOTE       = 0b10000;
    public static final int TAKE_STRING_NO_TRANSLATE_ESCAPES = 0b100000;

    public String takeString() {
        return takeString(TAKE_STRING_LEINIENT, "\"");
    }

    @Deprecated
    public String takeString(boolean leinient) {
        return takeString((leinient ? TAKE_STRING_LEINIENT : 0), "\"");
    }

    @Deprecated
    public String takeString(boolean leinient, boolean escapeDoubleQuote) {
        return takeString( (leinient ? TAKE_STRING_LEINIENT : 0) | (escapeDoubleQuote ? TAKE_STRING_ESCAPE_DOUBLE_QUOTE : 0), "\"");
    }

    @Deprecated
    public String takeString(boolean leinient, boolean escapeDoubleQuote, char quote) {
        return takeString((leinient ? TAKE_STRING_LEINIENT : 0) | (escapeDoubleQuote ? TAKE_STRING_ESCAPE_DOUBLE_QUOTE : 0), String.valueOf(quote));
    }

    public String takeString(int flags, String quote) {
        if ((flags & TAKE_STRING_NO_START_QUOTE) == 0) expect(quote);
        StringBuilder sb = new StringBuilder();
        while (!exhausted()) {
            String[] lines = takeUntil(quote).split("\r?\n", -1);
            String last = lines[lines.length - 1];
            if ((flags & (TAKE_STRING_MULTILINE | TAKE_STRING_ESCAPE_NEWLINE)) != 0) {
                for (int i = 0; i < lines.length - 1; i++) {
                    String next = lines[i];
                    if ((flags & TAKE_STRING_ESCAPE_NEWLINE) != 0) {
                        // count ending \'s
                        int count = 0;
                        for (int j = next.length() - 1; j >= 0; j--) {
                            if (next.charAt(j) == '\\') {
                                count++;
                            } else {
                                break;
                            }
                        }
                        if (count % 2 != 0) {
                            // escaped
                            if ((flags & TAKE_STRING_NO_TRANSLATE_ESCAPES) == 0) {
                                next = next.substring(0, next.length() - 1);
                            }
                            sb.append(next);
                            if ((flags & TAKE_STRING_NO_TRANSLATE_ESCAPES) != 0) {
                                sb.append('\n');
                            }
                        } else if (((flags & TAKE_STRING_MULTILINE) != 0)) {
                            sb.append(next).append('\n');
                        } else {
                            throw createException("Unexpected EOL in string literal");
                        }
                    } else {
                        sb.append(next).append('\n');
                    }
                }
            } else if (lines.length > 1) {
                throw createException("Unexpected EOL in string literal");
            }

            // count ending \'s
            int count = 0;
            for (int i = last.length() - 1; i >= 0; i--) {
                if (last.charAt(i) == '\\') {
                    count++;
                } else {
                    break;
                }
            }
            if (count % 2 != 0) {
                // escaped
                if ((flags & TAKE_STRING_NO_TRANSLATE_ESCAPES) == 0) {
                    last = last.substring(0, last.length() - 1);
                }
                sb.append(last);
                sb.append(expect(quote));
            } else {
                sb.append(last);
                if ((flags & TAKE_STRING_ESCAPE_DOUBLE_QUOTE) != 0) {
                    mark();
                    expect(quote);
                    try {
                        sb.append(expect(quote));
                        if ((flags & TAKE_STRING_NO_TRANSLATE_ESCAPES) != 0) {
                            sb.append(quote);
                        }
                    } catch (ParseException e) {
                        reset();
                        break;
                    }
                } else {
                    break;
                }
            }
        }
        expect(quote);
        if ((flags & TAKE_STRING_NO_TRANSLATE_ESCAPES) != 0) return quote + sb + quote;
        return StringUtils.translateEscapes(sb.toString(), (flags & TAKE_STRING_LEINIENT) != 0);
    }

    /* Char Acceptors */

    public char expect(char character) {
        return expect(character, false);
    }

    public char expect(char character, boolean ignoreCase) {
        int next = take();
        if (next == -1) throw createException("Expected " + character + " but got EOS");
        if (ignoreCase) {
            if (Character.toLowerCase((char) next) != Character.toLowerCase(character)) {
                throw createException("Expected " + character + " but got " + (char) next);
            }
        } else {
            if ((char) next != character) {
                throw createException("Expected " + character + " but got " + (char) next);
            }
        }
        return (char) next;
    }

    public char expect(String valueType, CharAccepter accepter) {
        int next = take();
        if (next == -1) throw createException("Expected " + valueType + " but got EOS");
        if (!accepter.accept((char) next)) {
            throw createException("Expected " + valueType + " but got " + (char) next);
        }
        return (char) next;
    }

    public String expect(String value) {
        return expect(value, false);
    }

    public String expect(String value, boolean ignoreCase) {
        for (char c : value.toCharArray()) {
            expect(c, ignoreCase);
        }
        return value;
    }

    public void expectEOF() {
        val c = take();
        if (c != -1) {
            throw createException("Expected EOF but got " + (char) c);
        }
    }

    public void expectEOS() {
        val c = take();
        if (c != -1) {
            throw createException("Expected EOS but got " + (char) c);
        }
    }

    /* CSV Specific */
    public List<String> takeRemainingCol() {
        return takeRemainingCol(true, COMMA);
    }

    public List<String> takeRemainingCol(boolean leinient) {
        return takeRemainingCol(leinient, COMMA);
    }

    public List<String> takeRemainingCol(boolean leinient, CharAccepter sep) {
        List<String> cols = new ArrayList<>();
        String next = takeCol(leinient, sep);
        while (next != null) {
            cols.add(next);
            next = takeCol(leinient, sep);
        }
        return cols;
    }

    public String takeCol() {
        return takeCol(true, COMMA);
    }

    public String takeCol(boolean leinient) {
        return takeCol(leinient, COMMA);
    }

    public @Nullable String takeCol(boolean leinient, CharAccepter sep) {
        int next = peek();
        if ((next == -1 || next == '\n')) {
            return null;
        }
        if (next == '"') {
            String value = takeString(TAKE_STRING_ESCAPE_DOUBLE_QUOTE | (leinient ? TAKE_STRING_LEINIENT : 0), "\"");
            String whiteSpace = takeNonNewlineWhitespace();
            next = peek();
            // check if string actually ends column
            if (next != '\n' && next != -1 && !sep.accept((char) next)) {
                if (!leinient) {
                    throw createException("Expected separator char, got " + next);
                }
                return value + whiteSpace + takeCol(true, sep);
            }
            // check if next is sep char and take it
            next = peek();
            if (next != -1 && sep.accept((char) next)) {
                take();
            }
            return value;
        }
        String value = takeUntil(or(sep, NEWLINE));
        next = peek();
        // check if next is sep char and take it
        if (sep.accept((char) next)) {
            take();
        }
        return value;
    }

    public <R> R parse(StringCharReader.ElementReader<R> reader) {
        this.mark();
        try {
            return reader.read(new WrappingReader(this, Integer.MAX_VALUE));
        } catch (ParseException t) {
            this.reset();
            throw t;
        }
    }

    @SafeVarargs
    public final <R> R parse(StringCharReader.ElementReader<R>... readers) {
        List<ParseException> exceptions = new ArrayList<>();
        for (StringCharReader.ElementReader<R> reader : readers) {
            try {
                return parse(reader);
            } catch (ParseException e) {
                exceptions.add(e);
            }
        }
        throw createCompositeException("Failed to parse as any", exceptions.toArray(new ParseException[0]));
    }

    @SafeVarargs
    public final <R> R parse(String type, StringCharReader.ElementReader<R>... readers) {
        List<ParseException> exceptions = new ArrayList<>();
        for (StringCharReader.ElementReader<R> reader : readers) {
            try {
                return parse(reader);
            } catch (ParseException e) {
                exceptions.add(e);
            }
        }
        throw createCompositeException("Failed to parse as any of " + type, exceptions.toArray(new ParseException[0]));
    }

    @Nullable
    @SafeVarargs
    public final <R> R parseOrNull(StringCharReader.ElementReader<R>... readers) {
        for (StringCharReader.ElementReader<R> reader : readers) {
            try {
                return parse(reader);
            } catch (ParseException ignored) {}
        }
        return null;
    }

    public ParseException createException(String message) {
        return createException(message, null);
    }

    public ParseException createException(String message, Throwable cause) {
        return new ParseException(message, cause);
    }

    public ParseException createCompositeException(String message, ParseException... exceptions) {
        val exception = createException(message);
        for (ParseException e : exceptions) {
            exception.addSuppressed(e);
        }
        return exception;
    }

    public interface ElementReader<R> {

        R read(CharReader<?> reader) throws ParseException;

    }

    public interface CharAccepter {
        boolean accept(char c);
    }

    /**
     * This wrapping class uses the reader class as a delegate,
     * so don't do multithreaded stuff, or interleave using the outer reader
     * with the wrapped reader.
     */
    public static class WrappingReader extends CharReader<WrappingReader> {
        private final StringBuilder sb = new StringBuilder();
        private final CharReader<?> reader;
        private final int limit;
        private int position = 0;
        private int mark = 0;

        public WrappingReader(CharReader<?> reader, int limit) {
            this.reader = reader;
            this.limit = limit;
        }

        @Override
        public int peek() {
            if (position == limit) {
                return -1;
            }
            if (position >= sb.length()) {
                int next = reader.take();
                if (next == -1) return -1;
                sb.append((char) next);
                return next;
            }
            return sb.charAt(position);
        }

        @Override
        public int take() {
            if (position == limit) {
                return -1;
            }
            if (position >= sb.length()) {
                int next = reader.take();
                position++;
                if (next == -1) return -1;
                sb.append((char) next);
                return next;
            }
            return sb.charAt(position++);
        }

        @Override
        public WrappingReader copy() {
            return new WrappingReader(reader, limit - position);
        }

        @Override
        public WrappingReader copy(int limit) {
            return new WrappingReader(reader, limit);
        }

        @Override
        public void mark(int limit) {
            mark = position;
        }

        @Override
        public void reset() {
            position = mark;
        }

        public String getAllRead() {
            return sb.toString();
        }
    }

}

File: commons-kt-kt/src/commonMain/kotlin/xyz/wagyourtail/commonskt/reader/CharReader.kt
package xyz.wagyourtail.commonskt.reader

import xyz.wagyourtail.commonskt.utils.translateEscapes
import kotlin.math.min

abstract class CharReader<T : CharReader<T>> {

    companion object {
        operator fun invoke(value: String) = StringCharReader(value)
    }

    /**
     * @return either a char, or null for eof
     */
    abstract fun peek(): Char?

    /**
     * @return either a char, or null for eof
     */
    abstract fun take(): Char?

    /**
     * @return the number of chars skipped
     */
    fun skip(count: Int): Int {
        var skipped = 0
        while (skipped < count) {
            val ch = take()
            if (ch == null) break
            skipped++
        }
        return skipped
    }

    /**
     * @return a string of the next count chars
     */
    open fun take(count: Int) = buildString {
        for (i in 0 until count) {
            take()?.let { append(it) } ?: break
        }
    }

    /**
     * @return a copy of the reader at the current position.
     * @since 1.0.4
     */
    abstract fun copy(limit: Int = Int.MAX_VALUE): CharReader<*>


    /**
     * set this position as remembered.
     */
    abstract fun mark(limit: Int = Int.MAX_VALUE)

    /**
     * reset to remembered mark position, or beginning if no mark.
     */
    abstract fun reset()

    fun exhausted() = peek() == null

    open fun takeRemaining() = buildString {
        while (peek() != null) append(take())
    }

    fun takeLine() = takeUntil('\n')

    open fun takeUntil(char: Char) = takeUntil { it == char }

    fun takeUntil(characters: String): String {
        val sb = StringBuilder()
        var next = peek()
        var taken = ""
        while (next != null) {
            do {
                if (characters.startsWith(taken + next)) {
                    if (taken.isEmpty()) {
                        mark()
                    }
                    taken += next
                    break
                } else if (!taken.isEmpty()) {
                    val newStart = taken.indexOf(characters[0], 1)
                    if (newStart == -1) {
                        taken = ""
                        if (characters.startsWith(taken + next)) {
                            mark()
                            taken += next
                        }
                        break
                    }
                    reset()
                    skip(newStart)
                    mark()
                    taken = taken.substring(newStart)
                    skip(taken.length)
                }
            } while (!taken.isEmpty())
            sb.append(take() as Char)
            if (characters == taken) {
                val out = sb.removeSuffix(characters)
                reset()
                return out.toString()
            }
            next = peek()
        }
        return sb.toString()
    }

    inline fun takeUntil(sep: (Char) -> Boolean) = buildString {
        var next = peek()
        while (next != null && !sep(next)) {
            append(take()!!)
            next = peek()
        }
    }

    inline fun takeWhile(acceptor: (Char) -> Boolean) = takeUntil { !acceptor(it) }

    fun takeWhitespace() = takeWhile { it.isWhitespace() }

    fun takeNonNewlineWhitespace() = takeWhile { it.isWhitespace() && it != '\n' }

    inline fun takeNext(sep: (Char) -> Boolean = { it.isWhitespace() }): String? {
        var next = peek()
        if (next == null || next == '\n') return null
        val value = if (next == '"') takeString()
        else takeUntil { sep(it) || it == '\n' }
        // take trailing sep
        next = peek()
        if (next != null && next != '\n' && sep(next)) {
            take()
        }
        return value
    }

    fun takeNext(sep: Char) = takeNext { it == sep }

    inline fun takeNextLiteral(sep: (Char) -> Boolean = { it.isWhitespace() }): String? {
        var next = peek()
        if (next == null || next == '\n') return null
        if (next == '\n') return null
        val value = takeUntil { sep(it) || it == '\n' }
        // take trailing sep
        next = peek()
        if (next != null && next != '\n' && sep(next)) {
            take()
        }
        return value
    }

    fun takeNextLiteral(sep: Char) = takeNextLiteral { it == sep }

    inline fun takeRemainingOnLine(sep: (Char) -> Boolean = { it.isWhitespace() }) = buildList<String> {
        var next = takeNext(sep)
        while (next != null) {
            add(next)
            next = takeNext(sep)
        }
    }

    fun takeRemainingOnLine(sep: Char) = takeRemainingOnLine { it == sep }

    inline fun takeRemainingLiteralOnLine(sep: (Char) -> Boolean = { it.isWhitespace() }) = buildList<String> {
        var next = takeNextLiteral(sep)
        while (next != null) {
            add(next)
            next = takeNextLiteral(sep)
        }
    }

    fun takeRemainingLiteralOnLine(sep: Char) = takeRemainingLiteralOnLine { it == sep }

//    fun takeWholeNumber(leadingZero: Boolean = false): Int {
//        if (leadingZero) {
//            return takeWhile { it.isDigit() }.toInt()
//        }
//    }

    fun takeString(leinient: Boolean = true, escapeDoubleQuote: Boolean = false, quote: Char = '"'): String {
        return takeString(leinient, escapeDoubleQuote, quote = quote.toString())
    }

    fun takeString(
        leinient: Boolean = true,
        escapeDoubleQuote: Boolean = false,
        escapeNewline: Boolean = false,
        multiline: Boolean = false,
        noStartQuote: Boolean = false,
        noTranslateEscapes: Boolean = false,
        quote: String = "\""
    ): String {
        if (!noStartQuote) {
            expect(quote)
        }
        val sb = StringBuilder()
        while (!exhausted()) {
            val lines = takeUntil(quote).split(Regex("\r?\n"))
            val last = lines.last()
            if (multiline || escapeNewline) {
                for (next in lines.dropLast(1)) {
                    if (escapeNewline) {
                        var count = 0
                        for (j in next.indices.reversed()) {
                            if (next[j] == '\\') {
                                count++
                            } else {
                                break
                            }
                        }
                        if (count % 2 != 0) {
                            // escaped
                            sb.append(
                                if (noTranslateEscapes) {
                                    "$next\n"
                                } else {
                                    next.substring(0, next.length - 1)
                                }
                            )
                        } else if (multiline) {
                            sb.append(next).append('\n')
                        } else {
                            throw createException("Unexpected EOL in string literal")
                        }
                    } else {
                        sb.append(next).append('\n')
                    }
                }
            } else if (lines.size > 1) {
                throw createException("Unexpected EOL in string literal")
            }

            var count = 0
            for (j in last.indices.reversed()) {
                if (last[j] == '\\') {
                    count++
                } else {
                    break
                }
            }
            if (count % 2 != 0) {
                // escaped
                sb.append(
                    if (!noTranslateEscapes) {
                        last.substring(0, last.length - 1)
                    } else {
                        last
                    }
                )
                sb.append(expect(quote))
            } else {
                sb.append(last)
                if (escapeDoubleQuote) {
                    mark()
                    expect(quote)
                    try {
                        sb.append(expect(quote))
                        if (noTranslateEscapes) {
                            sb.append(quote)
                        }
                    } catch (_: ParseException) {
                        reset()
                        break
                    }
                } else {
                    break
                }
            }
        }
        expect(quote)
        if (noTranslateEscapes) {
            return "$quote$sb$quote"
        }
        return sb.toString().translateEscapes(leinient)
    }

    fun expect(char: Char, ignoreCase: Boolean = false): Char {
        val it = take()
        if (it?.equals(char, ignoreCase) != true) {
            throw createException("Expected $char but got ${it ?: "EOS"}")
        }
        return char
    }

    fun expect(valueType: String, acceptor: (Char) -> Boolean): Char {
        val next = take() ?: throw createException("Expected $valueType but got EOS")
        if (!acceptor(next)) throw createException("Expected $valueType but got $next")
        return next
    }

    fun expect(value: String, ignoreCase: Boolean = false): String {
        for (char in value) {
            expect(char, ignoreCase)
        }
        return value
    }

    fun expectEOF() {
        val it = take()
        if (it != null) throw createException("Expected EOF but got $it")
    }

    fun expectEOS() {
        val it = take()
        if (it != null) throw createException("Expected EOS but got $it")
    }

    fun takeCol(leinient: Boolean = true, sep: (Char) -> Boolean = { it == ',' }): String? {
        var next = peek()
        if (next == null || next == '\n') {
            return null
        }
        if (next == '"') {
            val value = takeString(leinient, true)
            val whiteSpace = takeNonNewlineWhitespace()
            next = peek()
            if (next != null && next != '\n' && !sep(next)) {
                if (!leinient) {
                    throw createException("Expected separator char, got $next")
                }
                return value + whiteSpace + takeCol(true, sep)
            }
            next = peek()
            if (next != null && sep(next)) {
                take()
            }
            return value
        }
        val value = takeUntil { sep(it) || it == '\n' }
        next = peek()
        if (next != null && sep(next)) {
            take()
        }
        return value
    }

    fun takeCol(leinient: Boolean = true, sep: Char) = takeCol(leinient) { it == sep }

    fun takeRemainingCol(leinient: Boolean = true, sep: (Char) -> Boolean = { it == ',' }) = buildList<String> {
        var next = takeCol(leinient, sep)
        while (next != null) {
            add(next)
            next = takeCol(leinient, sep)
        }
    }

    fun takeRemainingCol(leinient: Boolean = true, sep: Char) = takeRemainingCol(leinient) { it == sep }

    open fun <R> parse(reader: CharReader<*>.() -> R): R {
        this.mark()
        try {
            val br = BufferedCharReader(this, Int.MAX_VALUE)
            val ret = reader(br)
            this.reset()
            this.skip(br.position)
            return ret
        } catch (e: ParseException) {
            this.reset()
            throw e
        }
    }

    fun <R> parse(vararg readers: CharReader<*>.() -> R): R {
        val exceptions = mutableListOf<ParseException>()
        for (reader in readers) {
            try {
                return parse(reader)
            } catch (e: ParseException) {
                exceptions.add(e)
            }
        }
        throw createCompositeException("Failed to parse", *exceptions.toTypedArray())
    }

    fun <R> parse(type: String, vararg readers: CharReader<*>.() -> R): R {
        val exceptions = mutableListOf<ParseException>()
        for (reader in readers) {
            try {
                return parse(reader)
            } catch (e: ParseException) {
                exceptions.add(e)
            }
        }
        throw createCompositeException("Failed to parse as any of $type", *exceptions.toTypedArray())
    }

    fun <R> parseOrNull(vararg readers: CharReader<*>.() -> R): R? {
        for (reader in readers) {
            try {
                return parse(reader)
            } catch (_: ParseException) {
            }
        }
        return null
    }

    open fun createException(msg: String, cause: Throwable? = null) = ParseException(msg, cause = cause)

    open fun createCompositeException(msg: String, vararg exceptions: ParseException) =
        createException(msg, null).also {
            for (e in exceptions) {
                it.addSuppressed(e)
            }
        }

    /**
     * This wrapping class uses the reader class as a delegate,
     * so don't do multithreaded stuff or interleave using the outer reader
     * with the wrapped reader.
     */
    class BufferedCharReader(private val reader: CharReader<*>, private var limit: Int) :
        CharReader<BufferedCharReader>() {
        private val buffer = StringBuilder()
        internal var position = 0
        private var mark = 0

        override fun peek(): Char? {
            return peekAt(position)
        }

        fun peekAt(position: Int): Char? {
            if (position == limit) {
                return null
            }
            if (position >= buffer.length) {
                return reader.peek()
            }
            return buffer[position]
        }

        override fun take(): Char? {
            return takeAt(position++)
        }

        fun takeAt(position: Int): Char? {
            if (position == limit) {
                return null
            }
            if (position == buffer.length) {
                val next = reader.take() ?: return null
                buffer.append(next)
                return next
            }
            return buffer[position]
        }

        override fun copy(limit: Int): CopyReader {
            if (limit < 0) throw IllegalArgumentException("limit < 0")
            val limit = position + limit
            return CopyReader(position, min(if (limit < position) Int.MAX_VALUE else limit, this.limit))
        }

        override fun <R> parse(reader: CharReader<*>.() -> R): R {
            this.mark()
            try {
                val br = copy()
                val ret = reader(br)
                this.position = br.position
                return ret
            } catch (e: ParseException) {
                this.reset()
                throw e
            }
        }

        override fun mark(limit: Int) {
            mark = position
        }

        override fun reset() {
            position = mark
        }

        inner class CopyReader(internal var position: Int, private val limit: Int) : CharReader<CopyReader>() {
            private var mark: Int = position

            override fun peek(): Char? {
                if (position == limit) {
                    return null
                }
                return peekAt(position)
            }

            override fun take(): Char? {
                if (position == limit) {
                    return null
                }
                return takeAt(position++)
            }

            override fun copy(limit: Int): CopyReader {
                if (limit < 0) throw IllegalArgumentException("limit < 0")
                val limit = position + limit
                return CopyReader(position, min(if (limit < position) Int.MAX_VALUE else limit, this.limit))
            }

            override fun <R> parse(reader: CharReader<*>.() -> R): R {
                this.mark()
                try {
                    val br = copy()
                    val ret = reader(br)
                    this.position = br.position
                    return ret
                } catch (e: ParseException) {
                    this.reset()
                    throw e
                }
            }

            override fun mark(limit: Int) {
                mark = position
            }

            override fun reset() {
                position = mark
            }

        }

    }

}
